relayer/src/supervisor.rs

```diff
@@ -0,0 +1,256 @@
+use std::{collections::HashMap, sync::Arc, thread::JoinHandle, time::Duration};
+
+use anomaly::BoxError;
+use crossbeam_channel::{Receiver, Sender};
+use itertools::Itertools;
+
+use ibc::{
+    events::IbcEvent,
+    ics04_channel::events::{CloseInit, SendPacket, TimeoutPacket, WriteAcknowledgement},
+    ics24_host::identifier::{ChainId, ChannelId, PortId},
+};
+
+use crate::{
+    chain::handle::ChainHandle,
+    event::monitor::EventBatch,
+    link::{Link, LinkParameters},
+};
+
+pub struct WorkerCmd {
+    pub batch: Arc<EventBatch>,
+}
+
+impl WorkerCmd {
+    pub fn new(batch: Arc<EventBatch>) -> Self {
+        Self { batch }
+    }
+}
+
+pub struct WorkerHandle {
+    pub tx: Sender<WorkerCmd>,
+    pub thread_handle: JoinHandle<()>,
+}
+
+impl WorkerHandle {
+    pub fn handle_packet_events(&self, batch: EventBatch) -> Result<(), BoxError> {
+        self.tx.send(WorkerCmd::new(Arc::new(batch)))?;
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+pub struct ChainHandlePair {
+    pub a: Box<dyn ChainHandle>,
+    pub b: Box<dyn ChainHandle>,
+}
+
+impl ChainHandlePair {
+    pub fn swap(self) -> Self {
+        Self {
+            a: self.b,
+            b: self.a,
+        }
+    }
+}
+
+pub struct Supervisor {
+    chains: ChainHandlePair,
+    workers: HashMap<Object, WorkerHandle>,
+}
+
+impl Supervisor {
+    pub fn spawn(
+        chain_a: Box<dyn ChainHandle>,
+        chain_b: Box<dyn ChainHandle>,
+    ) -> Result<Self, BoxError> {
+        let chains = ChainHandlePair {
+            a: chain_a,
+            b: chain_b,
+        };
+
+        Ok(Self {
+            chains,
+            workers: HashMap::new(),
+        })
+    }
+
+    pub fn run(mut self) -> Result<(), BoxError> {
+        let subscription_a = self.chains.a.subscribe()?;
+        let subscription_b = self.chains.b.subscribe()?;
+
+        loop {
+            println!("{} => iterating over event batches", self.chains.a.id());
```

Comment:
Could you remove this? And the one below?

Reply:
https://github.com/informalsystems/ibc-rs/pull/692/commits/c5f4da7f0af7773aab1da162cb81849729e743da